### ISSUE TITLE
Poor CPU utilization with Ray cluster: RaySwordfishActor only uses single core despite multi-core workers available

### ISSUE THREAD

**Issue by haifengwang1987** [Eventual-Inc/Daft#6126]
### Describe the bug

When running a Daft workload on a Ray cluster with multiple CPU cores available, the execution appears to be bottlenecked by a single CPU core. The `RaySwordfishActor` process consistently shows ~100% CPU usage (indicating single-core saturation), while other available CPU cores on the worker nodes remain idle.

### Environment

- **Daft version**: 0.7.2
- **Ray version**: 2.47.1
- **Python version**: 3.10.12
- **Operating System**: Ubuntu 22.04.2 LTS
- **Ray Cluster configuration**: 
  - 1 Head node (4 cores)
  - 1 Worker node (36 cores)

### Steps to Reproduce

I modified the official [Document Embedding benchmark](https://github.com/Eventual-Inc/Daft/tree/main/benchmarking/ai/document_embedding) with the following changes:

1. Downloaded all test data from S3 to local disk
2. Modified the code to read from local paths instead of S3
3. Removed the `Embedder` logic to simplify the workflow

**Modified code**:

```python
from __future__ import annotations

import pymupdf
import torch
import ray
from langchain_text_splitters import RecursiveCharacterTextSplitter
import time
import os
from datetime import datetime

import daft
from daft import col

PATH_PREFIX = "/mnt/data_nas"
EMBED_MODEL_ID = "sentence-transformers/all-MiniLM-L6-v2"
EMBED_MODEL_PATH = f"{PATH_PREFIX}/modelscope/hub/models/sentence-transformers/all-MiniLM-L6-v2"
EMBEDDING_DIM = 384
NUM_GPU_NODES = 3

now = datetime.now()
time_str = now.strftime('%Y%m%d%H%M%S')

INPUT_PATH = f"{PATH_PREFIX}/ray_example_data/digitalcorpora/metadata/**/"
OUTPUT_PATH = f"{PATH_PREFIX}/daft-benchmark-results/{time_str}/"
MAX_PDF_PAGES = 100
CHUNK_SIZE = 2048
CHUNK_OVERLAP = 200
EMBEDDING_BATCH_SIZE = 10

daft.set_runner_ray()

# Wait for Ray cluster to be ready
@ray.remote
def warmup():
    pass
ray.get([warmup.remote() for _ in range(64)])

def extract_text_from_parsed_pdf(pdf_bytes):
    try:
        doc = pymupdf.Document(stream=pdf_bytes, filetype="pdf")
        if len(doc) > MAX_PDF_PAGES:
            print(f"Skipping PDF because it has {len(doc)} pages")
            return None
        page_texts = [{"text": page.get_text(), "page_number": page.number} for page in doc]
        return page_texts
    except Exception as e:
        print(f"Error extracting text from PDF {e}")
        return None

def chunk(text):
    splitter = RecursiveCharacterTextSplitter(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP)
    chunk_iter = splitter.split_text(text)
    chunks = []
    for chunk_index, text in enumerate(chunk_iter):
        chunks.append({"text": text, "chunk_id": chunk_index})
    return chunks

def read_file_content(local_pdf_path):
    if os.path.exists(local_pdf_path):
        with open(local_pdf_path, "rb") as f:
            content = f.read()
            return content
    else:
        return None

start_time = time.time()
df = daft.read_parquet(INPUT_PATH)
df = df.where(daft.col("file_name").endswith(".pdf"))

df = df.with_column(
    "local_pdf_path",
    df["uploaded_pdf_path"].replace(
        "s3://ray-example-data/pdf_dump/",
        f"{PATH_PREFIX}/ray_example_data/digitalcorpora/",
    ),
)
df = df.with_column(
    "pdf_bytes",
    df["local_pdf_path"].apply(
        read_file_content,
        return_dtype=daft.DataType.binary(),
    ),
)

df = df.with_column(
    "pages",
    df["pdf_bytes"].apply(
        extract_text_from_parsed_pdf,
        return_dtype=daft.DataType.list(daft.DataType.struct({"text": daft.DataType.string(), "page_number": daft.DataType.int64()})),
    ),
)
df = df.explode("pages")
df = df.with_columns({"page_text": col("pages")["text"], "page_number": col("pages")["page_number"]})
df = df.where(daft.col("page_text").not_null())
df = df.with_column(
    "chunks",
    df["page_text"].apply(chunk, return_dtype=daft.DataType.list(daft.DataType.struct({"text": daft.DataType.string(), "chunk_id": daft.DataType.int64()})),
    ),
)
df = df.explode("chunks")
df = df.with_columns({"chunk": col("chunks")["text"], "chunk_id": col("chunks")["chunk_id"]})
df = df.where(daft.col("chunk").not_null())
df = df.select("uploaded_pdf_path", "page_number", "chunk_id")
df.write_parquet(OUTPUT_PATH)

end_time = time.time()
print(f"Time taken: {end_time - start_time} seconds")
```

### Expected Behavior
The workload should distribute across available CPU cores on the worker node, utilizing multiple cores in parallel to process different partitions/tasks.

### Actual Behavior
Only a single CPU core is fully utilized (RaySwordfishActor at ~102% CPU)
The remaining 35 cores on the worker node remain mostly idle
Performance is significantly lower than expected due to this bottleneck

### Evidence
1. CPU utilization from top command:

Initial execution showing RaySwordfishActor at 102.3% CPU:

<img width="863" height="223" alt="Image" src="https://github.com/user-attachments/assets/525ad279-7f2a-40ef-9c5e-dcaeb0bb10d4" />

Later during execution, still at ~102% CPU:

<img width="1040" height="268" alt="Image" src="https://github.com/user-attachments/assets/88eaa2ca-b011-489f-be58-7ed03041df4c" />

2. Ray Dashboard showing task execution:The Ray Dashboard shows that tasks are being executed, but the physical execution appears serialized:

<img width="1420" height="536" alt="Image" src="https://github.com/user-attachments/assets/589babca-77c5-49a8-a2b4-f3537c075259" />

### Questions
1. Is this single-core bottleneck expected behavior for CPU-based UDF operations (apply()) in Daft with Ray runner?
2. Are there configuration parameters (Daft or Ray) that could enable better parallelization across available cores?
3. Does the pipeline structure (multiple apply() + explode() operations) limit parallel execution?
4. Should I be setting specific resource requirements or concurrency limits to improve CPU utilization?

### Component(s)

UDFs (User Defined Functions)

### Additional context

The benchmark completes successfully but performance is severely limited by single-core execution. I expected Daft's Ray runner to automatically parallelize operations like apply(), explode(), and file I/O across available worker cores.

Note: The original benchmark with GPU-based Embedder had @daft.cls(max_concurrency=NUM_GPU_NODES, gpus=1) decorator, which was removed in this CPU-only version. Perhaps there's a CPU equivalent configuration needed?Any guidance on how to improve CPU utilization would be greatly appreciated!

**Comment by Jay-ju** [comment]
@haifengwang1987  I would like to ask, how many parquet files are there under your current `INPUT_PATH`?

**Comment by haifengwang1987** [comment]
The test data was downloaded from RayData's S3 bucket `s3://ray-example-data/digitalcorpora/metadata`, which contains 100 directories in total, with one parquet file in each directory, totaling 100 parquet files.

<img width="1278" height="161" alt="Image" src="https://github.com/user-attachments/assets/01c6ae8b-0dc0-4246-9e00-24e26dc72358" />

<img width="409" height="364" alt="Image" src="https://github.com/user-attachments/assets/198875dd-0781-4848-aa0d-e89ec3101588" />

**Comment by Jay-ju** [comment]
```
df = daft.read_parquet(INPUT_PATH)
df = df.where(daft.col("file_name").endswith(".pdf"))

# add paritions
df.explain(show_all=True)
```


Check here how many ScanTasks are finally generated; this is the total number of tasks. 
By default, one file corresponds to one task. If you think the concurrency is insufficient, you can test it by referring to https://docs.daft.ai/en/stable/api/config/#daft.context.set_execution_config to see if concurrency can be increased. Currently, the scan and downstream operators have the same morsel_size. If the concurrency after the scan needs to be adjusted, you can use into_batches to adjust it.

**Comment by haifengwang1987** [comment]
@Jay-ju Thanks for your reply.

This is the explain result of my task: [Explain Result](https://github.com/user-attachments/files/25096437/1.txt)

I tried adjusting the concurrency using `into_batches` and `repartitions`, but it didn't have any effect, it still only used the resources of a single CPU.

<img width="1264" height="148" alt="Image" src="https://github.com/user-attachments/assets/b6971e56-bf8a-4abc-8dd4-52ba64d47a4e" />

```python
df = df.with_column(
    "local_pdf_path",
    df["uploaded_pdf_path"].replace(
        "s3://ray-example-data/pdf_dump/",
        f"{PATH_PREFIX}/ray_example_data/digitalcorpora/",
    ),
)
df = df.into_batches(500)
df = df.with_column(
    "pdf_bytes",
    df["local_pdf_path"].apply(
        read_file_content,
        return_dtype=daft.DataType.binary(),
    ),
)
```
This is the explain result of adjusted task: [Adjusted Explain Result](https://github.com/user-attachments/files/25096902/3.txt)

<img width="1500" height="165" alt="Image" src="https://github.com/user-attachments/assets/f234b315-7275-48ca-8f71-cf6a92b62aa9" />

**Comment by Jay-ju** [comment]
@haifengwang1987 thank you for your reply. Have you tried the `_multithreaded_io` parameter?

**Comment by Jay-ju** [comment]
@haifengwang1987 What is the result like?

**Comment by haifengwang1987** [comment]
@Jay-ju Sorry for the late reply.

I tried setting the `_multithreaded_io` parameter in `read_parquet` to `True`, but it didn't work; it still only uses a single CPU. 
Later, I found a `use_process` parameter in `daft.func` in the Daft code. Setting this parameter to `True` allows the use of multiple CPU resources. 
Currently, I suspect that Python's GIL (Global Interpreter Lock) is causing the problem.

However, setting `use_process` to `True` revealed another problem: it causes task deadlocks. I will provide more complete information later.

**Comment by haifengwang1987** [comment]
When the use_process parameter of @daft.func is set to True, RaySwordfishActor will start multiple child processes to process data. However, it seems that udf.py will only read the standard output of the child process after the child process has finished processing. If the standard output or standard error logs of the child process are large, the output buffer may become full, causing the child process to be unable to write logs. At this time, the child process is blocked and and the task deadlock cannot end.

<img width="1886" height="238" alt="Image" src="https://github.com/user-attachments/assets/2a917d9b-08fe-4fa2-82a4-0c42749ed299" />

<img width="1620" height="80" alt="Image" src="https://github.com/user-attachments/assets/23c84500-5e7e-4817-be1f-a566af2d844e" />

<img width="1816" height="586" alt="Image" src="https://github.com/user-attachments/assets/cab011d0-0b00-48e5-b1e6-fb85cf9a4468" />

The `eval_input` function in `udf.py` can cause a deadlock.

```python
def eval_input(self, input: PyRecordBatch):
    # 1. send data
    self.handle_conn.send((shm_name, shm_size))
    
    # 2. Wait for a response (blocking)
    response = self.handle_conn.recv()  # If the child process is blocked due to stdout being full, it will never wait here
    
    # 3. Then read stdout
    stdout = self.trace_output()  # Too late!
```

Below is the deadlock stack trace captured by py-spy.

```
Process 2446031: /usr/bin/python -m daft.execution.udf_worker /tmp/tmpo7fgp0u1 9c1aa1d285f06d011d91678674acc0fa4ed0ddde356183db108fda6d5ef73aca
Python v3.10.12 (/usr/bin/python3.10)
Parent Process 2444477: ray::RaySwordfishActor

Thread 2446031 (idle): "MainThread"
    message (/usr/local/lib/python3.10/dist-packages/pymupdf/__init__.py:271)
        Arguments:
            text: "MuPDF error: format error: object out of range (63 0 R); xref size 5\n"
    JM_mupdf_error (/usr/local/lib/python3.10/dist-packages/pymupdf/__init__.py:20964)
        Arguments:
            text: "format error: object out of range (63 0 R); xref size 5"
    _print (/usr/local/lib/python3.10/dist-packages/pymupdf/mupdf.py:62759)
        Arguments:
            self: <DiagnosticCallbackPython at 0x7f26881bdbd0>
            message: "format error: object out of range (63 0 R); xref size 5"
    fz_load_page (/usr/local/lib/python3.10/dist-packages/pymupdf/mupdf.py:46912)
        Arguments:
            doc: <FzDocument at 0x7f26881bfe80>
            number: 7
    load_page (/usr/local/lib/python3.10/dist-packages/pymupdf/__init__.py:5786)
        Arguments:
            self: <Document at 0x7f26881bb400>
            page_id: 7
    __getitem__ (/usr/local/lib/python3.10/dist-packages/pymupdf/__init__.py:2912)
        Arguments:
            i: 7
    <listcomp> (/mnt/data_nas/code/daft_func.py:112)
        Arguments:
            .0: <iterator at 0x7f26881bfee0>
        Locals:
            page: <Page at 0x7f26881bfd00>
    extract_text_from_parsed_pdf_batch (/mnt/data_nas/code/daft_func.py:112)
        Arguments:
            pdf_bytes_series: <Series at 0x7f268b173400>
        Locals:
            pdf_bytes_list: [<bytes at 0x564802b79ae0>, <bytes at 0x564802be27d0>, <bytes at 0x564802c1dc20>, <bytes at 0x564802cdf490>, ...]
            results: [...]
            pdf_bytes: <bytes at 0x5648028d92f0>
            doc: <Document at 0x7f26881bb400>
            page_texts: [{"text": "abcd", "page_number": 0}, {"text": "abcd", "page_number": 1}, {"text": "abcd", "page_number": 2}, {...}, ...]
    method (/usr/local/lib/python3.10/dist-packages/daft/udf/udf_v2.py:92)
        Arguments:
            _self: None
        Locals:
            args: (<Series at 0x7f268b173400>)
            kwargs: {}
    bound_method (/usr/local/lib/python3.10/dist-packages/daft/udf/udf_v2.py:340)
        Locals:
            args: (<Series at 0x7f268b173400>)
            kwargs: {}
    call_batch_func (/usr/local/lib/python3.10/dist-packages/daft/udf/execution.py:86)
        Arguments:
            cls: <NoopCls at 0x7f2690108b80>
            method: <function at 0x7f268b4f3490>
            original_args: ((<Expression at 0x7f268b326890>), {})
            evaluated_args: [<builtins.PySeries at 0x7f268b173090>]
        Locals:
            evaluated_args_series: []
            args: (<Series at 0x7f268b173400>)
            kwargs: {}
            bound_method: <function at 0x7f2698b47eb0>
            metrics: <daft.daft.OperatorMetrics at 0x7f2698aadfc0>
    eval_expression_list_with_metrics (/usr/local/lib/python3.10/dist-packages/daft/recordbatch/recordbatch.py:247)
        Arguments:
            self: <RecordBatch at 0x7f268b37eb60>
            exprs: <ExpressionsProjection at 0x7f268b37dff0>
        Locals:
            pyexprs: [<daft.daft.PyExpr at 0x7f26993fda70>]
    udf_event_loop (/usr/local/lib/python3.10/dist-packages/daft/execution/udf_worker.py:60)
        Arguments:
            secret: <bytes at 0x7f268b4aa2e0>
            socket_path: "/tmp/tmpo7fgp0u1"
        Locals:
            conn: <Connection at 0x7f268b324c40>
            name: "psm_6864425e"
            expr_projection_bytes: <bytes at 0x564802846ad0>
            transport: <SharedMemoryTransport at 0x7f268b326830>
            expression_projection: <ExpressionsProjection at 0x7f268b37dff0>
            size: 24298760
            uninitialized_projection: <ExpressionsProjection at 0x7f268b326920>
            input_bytes: <bytes at 0x7f26805a6010>
            input: <RecordBatch at 0x7f268b37eb60>
    <module> (/usr/local/lib/python3.10/dist-packages/daft/execution/udf_worker.py:119)
    _run_code (/usr/lib/python3.10/runpy.py:86)
        Arguments:
            code: <code at 0x7f268b4b5dc0>
            run_globals: {"__name__": "__main__", "__doc__": None, "__package__": "daft.execution", ...}
            init_globals: None
            mod_name: "__main__"
            mod_spec: <ModuleSpec at 0x7f26991ee740>
            pkg_name: "daft.execution"
            script_name: None
        Locals:
            loader: <SourceFileLoader at 0x7f26991ed240>
            fname: "/usr/local/lib/python3.10/dist-packages/daft/execution/udf_worker.py"
            cached: "/usr/local/lib/python3.10/dist-packages/daft/execution/__pycache__/udf_worker.cpython-310.pyc"
    _run_module_as_main (/usr/lib/python3.10/runpy.py:196)
        Arguments:
            mod_name: "daft.execution.udf_worker"
            alter_argv: True
        Locals:
            mod_spec: <ModuleSpec at 0x7f26991ee740>
            code: <code at 0x7f268b4b5dc0>
            main_globals: {"__name__": "__main__", "__doc__": None, "__package__": "daft.execution", ...}

```

**Comment by Jay-ju** [comment]
@haifengwang1987  It still seems necessary to transfer the task to Ray's actor mode. Let's remove use_process. It also looks like we need into_batches to change the concurrency. You can try running into_batches(1) in an extreme case. I'm not sure about the concurrency from the source task to the execution of the UDF.

**Comment by desmondcheongzx** [comment]
Hey @haifengwang1987, really sorry for the delayed response, this issue ended up getting buried. Your py-spy work and the diagnosis around use_process=True are great and are super helpful!

On the single-core issue: the framing is more specific than "Flotilla only uses one core." Flotilla's Rust side does use all 36 cores on each worker - swordfish's tokio runtime spans the node, and Rust-native operators (scan, compute, shuffles) run across all of them. What's likely the bottleneck is the Python interpreter: we run one RaySwordfishActor per node, so there's one interpreter shared across all UDF work, and libraries like pymupdf and the langchain splitter don't release the GIL. So `top` shows ~100% on the one process because Python UDF work is GIL-serialized, regardless of `into_batches` or `_multithreaded_io` as those don't change what runs inside the interpreter.

The intended escape hatch is what you already found: `@daft.func(use_process=True)` (or `@daft.cls` if you have expensive per-worker init like a loaded model). Those spawn per-worker subprocesses with separate interpreters and will actually saturate the cores. We should have this flagged much more prominently in the CPU-UDF docs - sorry you had to find it the hard way.

On the deadlock: you're absolutely right. I've spun it out as #6762 with a minimal repro and fix direction

In the meantime, if you want to keep using `use_process=True`, adding `pymupdf.TOOLS.mupdf_display_errors(False)` at the top of your UDF and wrapping the page iteration in a `try`/`except` should suppress enough stderr that you don't fill the pipe. Also worth trying on 0.7.9 - unrelated to this bug, but there have been execution-path improvements since 0.7.2.

Going to close this out since the GIL bottleneck has a working answer (`use_process=True`) and the deadlock that was blocking you from using it is now tracked in #6762. Once that fix lands you should have a clean path to saturating all 36 cores. Happy to reopen if you think we're missing something.
